src/tap/src/tapcell.cpp

```diff
@@ -57,17 +63,1091 @@ Tapcell::~Tapcell()
+          masterl = pickCornerMaster(isXCorner(llx, rows_above, rows_below), row_ori, cnrcap_nwin_master, cnrcap_nwout_master, endcap_master);
+          masterr = pickCornerMaster(isXCorner(urx, rows_above, rows_below), row_ori, cnrcap_nwin_master, cnrcap_nwout_master, endcap_master);
+        }
+      } else {
+        masterl = endcap_master;
+        masterr = masterl;
+      }
+
+      if ((masterl->getWidth()) > (urx - llx)) {
+        continue;
+      }
+
+      buildCell(block, masterl, row_ori, llx, lly, prefix);
+
+      int master_x = masterr->getWidth();
+      int master_y = masterr->getHeight();
+
+      int loc_2_x = urx - master_x;
+      int loc_2_y = ury - master_y;
+      if (llx == loc_2_x && lly == loc_2_y) {
+        logger_->warn(utl::TAP, 9, "row {} have enough space for only one endcap.", subrow->getName());
+        continue;
+      }
+
+      odb::dbOrientType right_ori = row_ori;
+      if (row_ori == odb::dbOrientType::MX ) {
+        right_ori = odb::dbOrientType::R180;
+      } else if ( row_ori == odb::dbOrientType::R0) {
+        right_ori = odb::dbOrientType::MY;
+      }
+      buildCell(block, masterr, right_ori, loc_2_x, loc_2_y, prefix);
+    }
+  }
+
+  int endcap_count = Tapcell::phy_idx - start_phy_idx;
+  logger_->info(utl::TAP, 4, "Endcaps inserted: {}", endcap_count);
+  return endcap_count;
+}
+
+bool Tapcell::isXInRow(int& x, std::vector<odb::dbRow*>& subrow) {
+  for(odb::dbRow* row : subrow) {
+    odb::Rect row_bb;
+    row->getBBox(row_bb);
+    if (x >= row_bb.xMin() && x <= row_bb.xMax()) {
+      return 1;
+    }
+  }
+  return 0;
+}
+
+bool Tapcell::isXCorner(int& x, std::vector<odb::dbRow*>& rows_above, std::vector<odb::dbRow*>& rows_below) {
+  int in_above = isXInRow(x, rows_above);
+  int in_below = isXInRow(x, rows_below);
+
+  if (in_above && !in_below) {
+    return -1;
+  }
+  if (!in_above && in_below) {
+    return 1;
+  }
+  return 0;
+}
+
+odb::dbMaster* Tapcell::pickCornerMaster(int top_bottom, odb::dbOrientType ori, odb::dbMaster* cnrcap_nwin_master, odb::dbMaster* cnrcap_nwout_master, odb::dbMaster* endcap_master) {
+  if (top_bottom == 1 ) {
+    if (ori == odb::dbOrientType::MX ) {
+      return cnrcap_nwin_master;
+    } else {
+      return cnrcap_nwout_master;
+    }
+  } else if ( top_bottom == -1 ) {
+    if ( ori == odb::dbOrientType::R0 ) {
+      return cnrcap_nwin_master;
+    } else {
+      return cnrcap_nwout_master;
+    }
+  } else {
+    return endcap_master;
+  }
+}
+
+int Tapcell::insertTapcells(std::vector<vector<odb::dbRow*>>& rows, std::string tapcell_master, int& dist, std::string prefix) {
+  int start_phy_idx = Tapcell::phy_idx;
+  std::vector<int> xs;
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  int tap_width;
+  int dist1;
+  int dist2;
+  int y;
+  int x_start;
+  int x_end;
+  int x;
+  int row_y;
+  int gaps_above_below;
+  odb::dbOrientType ori;
+  int offset;
+  int pitch; 
+  int site_x;
+  int llx;
+  int lly;
+  int urx;
+  int ury;
```

Comment:
This is C (tcl) style where you define the variables far from their usage.  In c++ please define them in the small scope nearest the point of first usage.